pkg/sql/catalog/descriptor.go

```diff
@@ -632,11 +632,21 @@ type TableDescriptor interface {
 	// It's only non-nil if IsView is true.
 	GetDependsOnTypes() []descpb.ID
 
-	// GetConstraintInfoWithLookup returns a summary of all constraints on the
+	// GetNonDropConstraintInfoWithLookup returns a summary of all constraints on the
 	// table using the provided function to fetch a TableDescriptor from an ID.
-	GetConstraintInfoWithLookup(fn TableLookupFn) (map[string]descpb.ConstraintDetail, error)
-	// GetConstraintInfo returns a summary of all constraints on the table.
-	GetConstraintInfo() (map[string]descpb.ConstraintDetail, error)
+	GetNonDropConstraintInfoWithLookup(fn TableLookupFn) (map[string]descpb.ConstraintDetail, error)
+	// GetAllConstraintInfoWithLookup is similar to GetActiveAndInactiveConstraintInfoWithLookup
+	// except that
+	//  - it also includes constraints that are being dropped;
+	//  - the return uses `descpb.ConstraintID` as unique identifier.
+	GetAllConstraintInfoWithLookup(fn TableLookupFn) (map[descpb.ConstraintID]descpb.ConstraintDetail, error)
+	// GetNonDropConstraintInfo returns a summary of all constraints on the table.
+	GetNonDropConstraintInfo() (map[string]descpb.ConstraintDetail, error)
+	// GetAllConstraintInfo is similar to GetActiveAndInactiveConstraint
+	// except that
+	//  - it also includes constraints that are being dropped;
+	//  - the return uses `descpb.ConstraintID` as unique identifier.
+	GetAllConstraintInfo() (map[descpb.ConstraintID]descpb.ConstraintDetail, error)
```

Comment:
The existing `GetAllConstraintInfo.*` methods only are the way they are because of historical reasons, they don't really fit into our vision for what the catalog descriptor interfaces should look like. 

I would far prefer you define a function in `catalog` which builds this map using the new slice-returning methods you added to `catalog.TableDescriptor` and a lookup fn and options to filter non-drop constraints; and replace existing calls to `GetAllConstraintInfo.*` to calls of this function. It might even make things simpler and avoid copy-pasting big blocks of code as you do for `collectAllConstraintInfo`.